default/python/AI/pathfinding.py

```diff
@@ -1,16 +1,137 @@
+import FreeOrionAI as foAI
+import freeOrionAIInterface as fo
+import universe_object
+
 from heapq import heappush, heappop
 from collections import namedtuple
 
-import freeOrionAIInterface as fo
 from AIDependencies import INVALID_ID
+from EnumsAI import MissionType
 from turn_state import state
-
 from common.configure_logging import convenience_function_references_for_logger
-(debug, info, warn, error, fatal) = convenience_function_references_for_logger(__name__)
+from freeorion_tools import cache_by_session_with_turnwise_update, chat_human, get_partial_visibility_turn
 
 
+(debug, info, warn, error, fatal) = convenience_function_references_for_logger(__name__)
+_DEBUG_CHAT = False
 _ACCEPTABLE_DETOUR_LENGTH = 2000
 
+# cache this so that boost python does not need to make a new copy every time this info is needed in a turn.
+@cache_by_session_with_turnwise_update
+def _get_unobstructed_systems():
+    return fo.getEmpire().supplyUnobstructedSystems
+
+def _info_string(path_info):
+    sequence_string = ", ".join([str(universe_object.System(sys_id)) for sys_id in path_info.path])
+    return "dist %.1f, path %s" % (path_info.distance, sequence_string)
+
+# Note that this can cover departure from uncontested systems as well as from contested systems where our forces
+# arrived first (and those forces may or may not be the forces for whom the pathfinding is being done)
+def _more_careful_travel_starlane_func(c, d):
+    return c in _get_unobstructed_systems()
+
+
+# In comparison to  _more_careful_travel_starlane_func, this can also allow passage through contested
+# systems where our forces already present there (which may or may not be the forces for whome the pathfinding
+# is being done) arrived second but are currently preserving travel along the starlane of interest.
+def _somewhat_careful_travel_starlane_func(c, d):
+    return any((c in _get_unobstructed_systems(),
+                fo.getEmpire().unrestrictedLaneTravel(c, d)))
+
+
+# For some activities like scouting, we may want to allow an extra bit of risk from routing through a system which
+# has all its exits necessarily marked as restricted simply because it has never been partially visible
+def _risky_travel_starlane_func(c, d):
+    return any((_somewhat_careful_travel_starlane_func(c, d),
+                get_partial_visibility_turn(c) <= 0))
+
+
+def _always_ok_travel_starlane_func(c, d):
+    return True
+
+
+def _always_ok_travel_system_func(s):
+    return True
+
+_STARLANE_TRAVEL_FUNC_MAP = {
+    MissionType.EXPLORATION: _risky_travel_starlane_func,
+    MissionType.OUTPOST: _more_careful_travel_starlane_func,
+    MissionType.COLONISATION: _more_careful_travel_starlane_func,
+    MissionType.INVASION: _somewhat_careful_travel_starlane_func,
+    MissionType.MILITARY: _always_ok_travel_starlane_func,
+    MissionType.SECURE: _always_ok_travel_starlane_func,
+}
```

Comment:
> I would like to see what the fleet is allowed to do or not just from looking at this table. The function names are completely nondescript.

Hmm, I have mixed thoughts on this, probably partly because I have the experience already of seeing the functions evolve a bit, and am contemplating how they might evolve more, and I don't really like having to change the function names just because we change the logic somewhat, and to me  the current names seem like a reasonable degree of abstraction, not completely nondescript, 

And once you start incorporating a significant amount of decision logic into the name then it seems overly subject to becoming either inaccurate or unwieldy.  For example, your proposed name ```may_enter_blockaded_systems_if_exploring``` seems overly inaccurate to me because it is actually just a subset of blockaded systems for which travel is allowed through (and that subset typically becomes a relatively very small subset), but trying to put the whole decision logic in the name seems unwieldy.

That said, I don't have any significant objection to your proposed names ```may_only_travel_unobstructed_systems``` and ```may_only_travel_unblockaded_starlanes```, though it would leave me wary of the need to change the names in the future, and I wouldn't be crazy about the significant name change style that I tend to suspect would remain for the function currently named ```_risky_travel_starlane_func``` even if we do change that to something else.

And consider also these thoughts for the future-- these functions all still allow, to some degree or another, travel through contested systems, and  we simply leave it to elsewhere for a risk averse fleet to decide if the next system on its route is risky enough that it wants to wait before entering.  Now, because of this new pathfinding development a lot of that other code can surely be more readily improved, but perhaps we'd also like to consider having some fleets try taking an even more risk-averse path that avoids contested systems altogether, but falls back to a simply unblockaded path if it can't find an uncontested one.  Now, ```may_only_travel_uncontested_systems``` would be a fine name for that within the naming scheme you are proposing, and this kind of more involved decision I am talking about would probably need to involve a lot more changes to this pathfinder module than simply adding a new travel starlane function or a new travel system function.

So... if you'd still like to proceed at this time with the name changes for ```may_only_travel_unobstructed_systems``` and ```may_only_travel_unblockaded_starlanes``` and either have a better replacement for ```_risky_travel_starlane_func``` or are fine with the naming style being different, then please just confirm that & I'll be OK with the change.

Reply:
> to me the current names seem like a reasonable degree of abstraction, not completely nondescript

I really do not like the current names and wouldn't agree with that sentiment but i can see how choosing a more precise name would depend on the planned usage and extension in the future.

If you have a plan how to progress from here and how you want to extend the logic and think you will be able to find the time to work on that "in the near future" (in FreeOrion time scale), then I am fine with leaving as is. If we are going to leave this unattended for some time, I would prefer to sort this out.